packages/core/content-releases/server/src/middlewares/documents.ts

```diff
@@ -0,0 +1,122 @@
+import { Modules, UID } from '@strapi/types';
+import { RELEASE_MODEL_UID, RELEASE_ACTION_MODEL_UID } from '../constants';
+import { getService, isEntryValid } from '../utils';
+
+type Middleware = Modules.Documents.Middleware.Middleware;
+
+interface ReleaseActionsParams {
+  contentType: UID.ContentType;
+  entryDocumentId?: Modules.Documents.ID;
+  locale?: string;
+}
+
+const updateActionsStatusAndUpdateReleaseStatus = async (
+  contentType: UID.ContentType,
+  entry: Modules.Documents.AnyDocument
+) => {
+  const releases = await strapi.db.query(RELEASE_MODEL_UID).findMany({
+    where: {
+      actions: {
+        contentType,
+        entryDocumentId: entry.documentId,
+        locale: entry.locale,
+      },
+    },
+  });
+
+  const entryStatus = await isEntryValid(contentType, entry, { strapi });
+
+  await strapi.db.query(RELEASE_ACTION_MODEL_UID).update({
+    where: {
+      contentType,
+      entryDocumentId: entry.documentId,
+      locale: entry.locale,
+    },
+    data: {
+      isEntryValid: entryStatus,
+    },
+  });
+
+  for (const release of releases) {
+    getService('release', { strapi }).updateReleaseStatus(release.id);
+  }
+};
+
+const deleteActionsAndUpdateReleaseStatus = async (params: ReleaseActionsParams) => {
+  const releases = await strapi.db.query(RELEASE_MODEL_UID).findMany({
+    where: {
+      actions: params,
+    },
+  });
+
+  await strapi.db.query(RELEASE_ACTION_MODEL_UID).deleteMany({
+    where: params,
+  });
+
+  for (const release of releases) {
+    getService('release', { strapi }).updateReleaseStatus(release.id);
+  }
+};
+
+const deleteActionsOnDelete: Middleware = async (ctx, next) => {
+  if (ctx.action !== 'delete') {
+    return next();
+  }
+
+  if (!ctx.contentType.options?.draftAndPublish) {
```

Comment:
could you use the following `@ strapi/utils ` utility to check this?
```suggestion
  import { contentTypes } from "@ strapi/utils"

  if (contentTypes.hasDraftAndPublish(ctx.contentType) {
```